mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -310,11 +470,25 @@ object StringIndexerModel extends MLReadable[StringIndexerModel] {
     override def load(path: String): StringIndexerModel = {
       val metadata = DefaultParamsReader.loadMetadata(path, sc, className)
       val dataPath = new Path(path, "data").toString
-      val data = sparkSession.read.parquet(dataPath)
-        .select("labels")
-        .head()
-      val labels = data.getAs[Seq[String]](0).toArray
-      val model = new StringIndexerModel(metadata.uid, labels)
+
+      // We support to load old `StringIndexerModel` saved by previous Spark versions.
+      // Previous model has `labels`, but new model has `labelsArray`.
+      val (majorVersion, minorVersion) = majorMinorVersion(metadata.sparkVersion)
+      val labelsArray = if (majorVersion < 3) {
+        // Spark 2.4 and before.
```

Comment:
Do we have test for this?

Reply:
This kind of test might need to commit a model from before this change. I'm not sure if there is similar tests in other ML models. Let me check about it.

Reply:
I committed a model generated in Spark 2.4. Then I added a test for loading this model.

Reply:
Don't we normally handle this with model versions rather than tie to Spark versions?

Reply:
There are also others using Spark versions to handle this, like Word2Vec, LogisticRegression.

Reply:
Hm, that's not great, but OK